kratos/includes/geometry_container.h

```diff
@@ -0,0 +1,407 @@
+        : Flags()
+        , mpGeometries(NewGeometries)
+    {}
+
+    /// Destructor
+    ~GeometryContainer() override
+    {}
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    GeometryContainer Clone()
+    {
+        typename GeometriesContainerType::Pointer p_geometries(new GeometriesContainerType(*mpGeometries));
+
+        return GeometryContainer(p_geometries);
+    }
+
+    void Clear()
+    {
+        Flags::Clear();
+        DataValueContainer::Clear();
+        mpGeometries->clear();
+    }
+
+    ///@}
+    ///@name Geometries
+    ///@{
+
+    /// Return number of geometries stored inside this geometry container
+    SizeType NumberOfGeometries() const
+    {
+        return mpGeometries->size();
+    }
+
+    ///@}
+    ///@name Add Functions
+    ///@{
+
+    /// Adds a geometry to the geometry container.
+    GeometryIterator AddGeometry(typename TGeometryType::Pointer pNewGeometry)
+    {
+        auto i = mpGeometries->find(pNewGeometry->Id());
+        if(i == mpGeometries->end())
+            return mpGeometries->insert(pNewGeometry);
+        else
+        {
+            KRATOS_ERROR << "Geometry with Id: " << pNewGeometry->Id()
+                << " exists already.";
+        }
+    }
+
+    ///@}
+    ///@name Get Functions
+    ///@{
+
+    /// Returns the Geometry::Pointer corresponding to its Id
+    typename TGeometryType::Pointer pGetGeometry(IndexType GeometryId)
+    {
+        auto i = mpGeometries->find(GeometryId);
+        KRATOS_ERROR_IF(i == mpGeometries->end())
+            << " geometry index not found: " << GeometryId << ".";
+        return (i.base()->second);
+    }
+
+    /// Returns the Geometry::Pointer corresponding to its name
+    typename TGeometryType::Pointer pGetGeometry(std::string GeometryName)
+    {
+        auto hash_index = TGeometryType::GenerateId(GeometryName);
+        auto i = mpGeometries->find(hash_index);
+        KRATOS_ERROR_IF(i == mpGeometries->end())
+            << " geometry index not found: " << GeometryName << ".";
+        return (i.base()->second);
+    }
+
+    /// Returns a reference geometry corresponding to the id
+    TGeometryType& GetGeometry(IndexType GeometryId)
+    {
+        auto i = mpGeometries->find(GeometryId);
+        KRATOS_ERROR_IF(i == mpGeometries->end()) << " geometry index not found: " << GeometryId << ".";
+        return *i;
+    }
+
+    /// Returns a const reference geometry corresponding to the name
+    const TGeometryType& GetGeometry(std::string GeometryName) const
+    {
+        auto hash_index = TGeometryType::GenerateId(GeometryName);
+        auto i = mpGeometries->find(hash_index);
+        KRATOS_ERROR_IF(i == mpGeometries->end())
+            << " geometry index not found: " << GeometryName << ".";
+        return *i;
+    }
+
+    ///@}
+    ///@name Remove Functions
+    ///@{
+
+    /// Remove the geometry with given Id from geometry container
+    void RemoveGeometry(IndexType GeometryId)
```

Comment:
maybe this one fct is enough, since from the other remove fcts you always call this one
not sure though, esp since I currently dont fully get what the GeometryName is

Reply:
They're all deleting with different accesses. I know that it could be avoided, because in the end only the id is relevant. However, I still am in favor keeping all. Can be discussed.

Reply:
We removed the redundant functions. Thanks for your comment!